### PR TITLE
Watch Subdirectories of Watched folders in Blocked Mods Dialog

### LINKED ISSUE #938: Mod Download Folder not recursively searched.
### Operating System

Windows

### Version of Prism Launcher

Prism Launcher 6.3

### Version of Qt

QT 5.15.2

### Description of bug

When adding a modpack, the searched folder for mods does not recursively search through sub folders.
Prism Search

### Steps to reproduce

Add mods in subfolder
Add modpack
Watch fail

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Ryex**
What is says on the tin. Pretty simple addition

closes #938

### CODE REVIEW

File: launcher/ui/dialogs/BlockedModsDialog.cpp [hunk 1]
```diff
@@ -186,8 +185,26 @@ void BlockedModsDialog::setupWatch()
...
+    m_watcher.addPath(path);
+
+    if (!watch_subdirectories)
+        return;
+
+    QDirIterator it(path, QDir::Filter::Dirs, QDirIterator::NoIteratorFlags);
+    while (it.hasNext()) {
+        QString dir_path = it.next();
+        QDir to_watch_dir(dir_path);
+        if (to_watch_dir.dirName() == "." || to_watch_dir.dirName() == "..")
```
Comment (flowln) [inline launcher/ui/dialogs/BlockedModsDialog.cpp]: any reason for not using `QDir::NoDotAndDotDot` in the filter flags up there?

File: launcher/ui/dialogs/BlockedModsDialog.cpp [hunk 2]
```diff
@@ -186,8 +185,26 @@ void BlockedModsDialog::setupWatch()
 {
     const QString downloadsFolder = APPLICATION->settings()->get("DownloadsDir").toString();
     const QString modsFolder = APPLICATION->settings()->get("CentralModsDir").toString();
-    m_watcher.addPath(downloadsFolder);
-    m_watcher.addPath(modsFolder);
+    watchPath(downloadsFolder, true);
```
Comment (flowln) [inline launcher/ui/dialogs/BlockedModsDialog.cpp]: i don't think this is good default behavior, i wouldn't really feel super well about my MC launcher looking for mods in my MuseScore macOS bundle, were I an average user

the recursive behavior causes a lot of completely unrelated folders inside the downloads folder to be watched as well

File: launcher/ui/dialogs/BlockedModsDialog.cpp (hunk 1 above)
Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: mostly that I forgot it exists

File: launcher/ui/dialogs/BlockedModsDialog.cpp (hunk 2 above)
Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: hmm, yeah. this is why I questioned the wisdom of it being default in the issue. I figured I put it out and get comments during review. I'll add a setting.

**Comment by Ryex** [comment]
Recursive search on the downloads folder is now attached to a setting.

File: launcher/ui/pages/global/LauncherPage.ui [hunk 3]
```diff
@@ -117,33 +123,37 @@
...
+             <cstring>iconsDirTextBox</cstring>
             </property>
            </widget>
           </item>
-          <item row="3" column="1">
-           <widget class="QLineEdit" name="downloadsDirTextBox"/>
-          </item>
-          <item row="3" column="2">
-           <widget class="QToolButton" name="downloadsDirBrowseBtn">
+          <item row="4" column="0">
```
Comment (flowln) [inline launcher/ui/pages/global/LauncherPage.ui]:
```suggestion
          <item row="4" column="1" colspan="2">
```
just to make it look a bit better

File: launcher/ui/pages/global/LauncherPage.ui [hunk 4]
```diff
@@ -117,33 +123,37 @@
...
+           <widget class="QCheckBox" name="downloadsDirWatchRecursiveCheckBox">
+            <property name="toolTip">
+             <string>when looking for mods in places like the blocked mods dialog Prismlauncher will check in sub folders of your downloads folder too.</string>
```
Comment (flowln) [inline launcher/ui/pages/global/LauncherPage.ui]:
```suggestion
             <string>When enabled, in addition to the downloads folder, its sub folders will also be searched when looking for resources (e.g. when looking for blocked mods on CurseForge).</string>
```

File: launcher/ui/pages/global/LauncherPage.cpp [hunk 5]
```diff
@@ -184,6 +184,11 @@ void LauncherPage::on_downloadsDirBrowseBtn_clicked()
     }
 }
 
+void LauncherPage::on_downloadsDirWatchRecursiveCheckBox_clicked()
+{
+  // incase anything needs to be done here
```
Comment (flowln) [inline launcher/ui/pages/global/LauncherPage.cpp]: if something needs to be done here in the future, we can add the method, there's no need to add it if we don't use it now.

File: launcher/ui/dialogs/BlockedModsDialog.cpp [hunk 6]
```diff
@@ -186,8 +185,24 @@ void BlockedModsDialog::setupWatch()
...
+    m_watcher.addPath(path);
+
+    if (!watch_subdirectories)
+        return;
+
+    QDirIterator it(path, QDir::Filter::Dirs | QDir::Filter::NoDotAndDotDot, QDirIterator::NoIteratorFlags);
+    while (it.hasNext()) {
+        QString dir_path = it.next();
+        watchPath(dir_path, watch_subdirectories);
+    }
```
Comment (flowln) [inline launcher/ui/dialogs/BlockedModsDialog.cpp]: i think this needs a bit of logic to not recurse into itself with symlinks
symbolic link in mods folder, linking to the mods folder itself, being followed recursively

Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: now that's just evil >:(

please confirm this is fixed with the latest :P

Comment (flowln) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: The recursion displayed before is fixed, though the `mods` folder now got added twice:
two mods folders are present: one with a relative path, and one with an absolute path
not a big issue though

Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: ok, now it *really* can't duplicate a path to watch!

Comment (flowln) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: now it doesn't work as intended though :P

Commit 5ce7874:
image

Commit 62c5982:
image

I suppose in the line `auto to_watch_path = to_watch.canonicalPath();`, the call to `canonicalPath()` needs to be replaced by `canonicalFilePath()` so as to not consume the last element of the path.

Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.cpp]: .....
*sigh*

clearly I forgot to rebuild somehow before I ran my tests.

File: launcher/ui/pages/global/LauncherPage.cpp
Comment (flowln) [inline launcher/ui/pages/global/LauncherPage.cpp:1]: you're missing setting the checkbox state in `loadSettings()`

File: launcher/ui/pages/global/LauncherPage.ui [hunk 7]
```diff
@@ -90,16 +87,25 @@
...
           <item row="2" column="1">
            <widget class="QLineEdit" name="iconsDirTextBox"/>
           </item>
-          <item row="2" column="2">
-           <widget class="QToolButton" name="iconsDirBrowseBtn">
+          <item row="3" column="2">
+           <widget class="QToolButton" name="downloadsDirBrowseBtn">
             <property name="text">
-             <string notr="true">...</string>
+             <string>...</string>
```
Comment (flowln) [inline launcher/ui/pages/global/LauncherPage.ui]:
```suggestion
             <string notr="true">...</string>
```

the other `...`s in this page are notr

File: launcher/ui/pages/global/LauncherPage.cpp
Comment (Ryex) [reply launcher/ui/pages/global/LauncherPage.cpp:1]: fixed

File: launcher/ui/pages/global/LauncherPage.ui (hunk 7 above)
Comment (Ryex) [reply launcher/ui/pages/global/LauncherPage.ui]: also fixed, seems it got left out in the previous work on this file.

File: launcher/ui/dialogs/BlockedModsDialog.h [hunk 8]
```diff
@@ -79,6 +79,7 @@ class BlockedModsDialog : public QDialog {
     void update();
     void directoryChanged(QString path);
     void setupWatch();
+    void watchPath(QString path, bool watch_subdirectories = false);
```
Comment (Scrumplex) [inline launcher/ui/dialogs/BlockedModsDialog.h]: Maybe call the parameter `recursive` instead. The current name could also mean that it doesn't descend more than once.

**Review by Scrumplex — APPROVED** [review]
LGTM. Works on my machine:tm:

**Review by DioEgizio — APPROVED** [review]
works on my machine

File: launcher/ui/dialogs/BlockedModsDialog.h (hunk 8 above)
Comment (Ryex) [reply launcher/ui/dialogs/BlockedModsDialog.h]: fixed :+1:

**Pull request merged**